server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMessagesMethod.java

```diff
@@ -0,0 +1,144 @@
+import org.apache.james.mailbox.store.mail.MailboxMapperFactory;
+import org.apache.james.mailbox.store.mail.MessageMapper;
+import org.apache.james.mailbox.store.mail.MessageMapper.FetchType;
+import org.apache.james.mailbox.store.mail.model.Mailbox;
+import org.apache.james.mailbox.store.mail.model.MailboxId;
+import org.apache.james.mailbox.store.mail.model.MailboxMessage;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class SetMessagesMethod<Id extends MailboxId> implements Method {
+
+    private static final int LIMIT_BY_ONE = 1;
+    private static final Logger LOGGER = LoggerFactory.getLogger(SetMessagesMethod.class);
+    private static final Method.Request.Name METHOD_NAME = Method.Request.name("setMessages");
+    private static final Method.Response.Name RESPONSE_NAME = Method.Response.name("messagesSet");
+
+    private final MailboxMapperFactory<Id> mailboxMapperFactory;
+    private final MailboxSessionMapperFactory<Id> mailboxSessionMapperFactory;
+    private final SetMessagesUpdateProcessor<Id> messageUpdater;
+
+    @Inject
+    @VisibleForTesting SetMessagesMethod(MailboxMapperFactory<Id> mailboxMapperFactory,
+                                         MailboxSessionMapperFactory<Id> mailboxSessionMapperFactory, SetMessagesUpdateProcessor<Id> messageUpdater) {
+        this.mailboxMapperFactory = mailboxMapperFactory;
+        this.mailboxSessionMapperFactory = mailboxSessionMapperFactory;
+        this.messageUpdater = messageUpdater;
+    }
+
+    @Override
+    public Method.Request.Name requestHandled() {
+        return METHOD_NAME;
+    }
+
+    @Override
+    public Class<? extends JmapRequest> requestType() {
+        return SetMessagesRequest.class;
+    }
+
+    public Stream<JmapResponse> process(JmapRequest request, ClientId clientId, MailboxSession mailboxSession) {
+        Preconditions.checkArgument(request instanceof SetMessagesRequest);
+        try {
+            return Stream.of(
+                    JmapResponse.builder().clientId(clientId)
+                    .response(setMessagesResponse((SetMessagesRequest) request, mailboxSession))
+                    .responseName(RESPONSE_NAME)
+                    .build());
+        } catch (MailboxException e) {
+            return Stream.of(
+                    JmapResponse.builder().clientId(clientId)
+                    .error()
+                    .responseName(RESPONSE_NAME)
+                    .build());
+        }
+    }
+
+    private SetMessagesResponse setMessagesResponse(SetMessagesRequest request, MailboxSession mailboxSession) throws MailboxException {
+        SetMessagesResponse.Builder responseBuilder = SetMessagesResponse.builder();
+        processDestroy(request.getDestroy(), mailboxSession, responseBuilder);
+        processUpdatesAndMergeResponse(request, mailboxSession, responseBuilder);
```

Comment:
extract to _SetMessageResponse_.**mergeInto**(_SetMessageResponse.Builder_ builde